### PR TITLE
[SPARK-41982][SQL] Partitions of type string should not be treated as numeric types

### PR DESCRIPTION
**Pull request by smallzhongfeng**
### What changes were proposed in this pull request?

Ensure that partitions of type string without quotation marks are not recognized as numeric types.
For example:
```
create table if not exists test_90(a string, b string) partitioned by (dt string);
desc formatted test_90;
insert into table test_90 partition (dt=05) values("1","2");
insert into table test_90 partition (dt='05') values("1","2");
drop table test_90;
```
before spark3.1 and earlier, it will generate such a path: `hdfs://test5/user/hive/db1/test_90/dt=05`

```
spark-sql> select * from test_90;
1       2       05
1       2       05
Time taken: 1.316 seconds, Fetched 2 row(s)

spark-sql> show partitions test_90; 
dt=05 
Time taken: 0.201 seconds, Fetched 1 row(s)

spark-sql> select * from test_90 where dt='05';
1       2       05
1       2       05
Time taken: 0.212 seconds, Fetched 2 row(s)
```

after spark3.1, it will generate two path:  `hdfs://test5/user/hive/db1/test_90/dt=05` and `hdfs://test5/user/hive/db1/test_90/dt=5`

```
spark-sql> select * from test_90;
1       2       05
1       2       5
Time taken: 2.119 seconds, Fetched 2 row(s)

spark-sql> show partitions test_90;
dt=05
dt=5
Time taken: 0.161 seconds, Fetched 2 row(s)

spark-sql> select * from test_90 where dt='05';
1       2       05
Time taken: 0.252 seconds, Fetched 1 row(s)
```

This will cause inconsistent read data. After seeing [https://github.com/apache/spark/pull/30421](https://github.com/apache/spark/pull/30421), I think if the user does not know about this change and the migration document does not mention it, I think it will affect the data quality, so I added the parameter `spark.sql.legacy.keepPartitionSpecAsStringLiteral`, which will maintain the original effect when the parameter set `true`.

### Why are the changes needed?

If the partition is of `String`, but the value of partition without quotation marks, it will still be treated as `String` through parameter configuration.

### Does this PR introduce _any_ user-facing change?

After the parameter `spark.sql.legacy.keepPartitionSpecAsStringLiteral` is enabled, the partition path generated by partition `partition (dt=05)` and partition `partition (dt='05')` is the same.

### How was this patch tested?

New uts.

### CODE REVIEW

**Comment by smallzhongfeng** [comment]
cc @AngersZhuuuu @cloud-fan @maropu Hope to get your reply!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 1]
```diff
@@ -507,14 +507,19 @@ class AstBuilder extends SqlBaseParserBaseVisitor[AnyRef] with SQLConfHelper wit
       ctx: PartitionSpecContext): Map[String, Option[String]] = withOrigin(ctx) {
     val legacyNullAsString =
       conf.getConf(SQLConf.LEGACY_PARSE_NULL_PARTITION_SPEC_AS_STRING_LITERAL)
+    val keepPartitionTypeAsString =
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:510]:
```suggestion
    val keepPartitionSpecAsString =
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 2]
```diff
@@ -4059,6 +4059,15 @@ object SQLConf {
       .booleanConf
       .createWithDefault(true)
 
+  val LEGACY_KEEP_PARTITION_SPEC_AS_STRING_LITERAL =
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:4062]: Can we put it near LEGACY_PARSE_NULL_PARTITION_SPEC_AS_STRING_LITERAL and follow its cofig doc as well?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala [hunk 3]
```diff
@@ -363,7 +363,8 @@ class SparkSqlAstBuilder extends AstBuilder {
    * Convert a constants list into a String sequence.
    */
   override def visitConstantList(ctx: ConstantListContext): Seq[String] = withOrigin(ctx) {
-    ctx.constant.asScala.map(v => visitStringConstant(v, legacyNullAsString = false)).toSeq
+    ctx.constant.asScala.map(v => visitStringConstant(v,
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala:366]: can we add default parameter value for `visitStringConstant`?

File: sql/core/src/test/scala/org/apache/spark/sql/SQLInsertTestSuite.scala [hunk 4]
```diff
@@ -351,6 +351,36 @@ trait SQLInsertTestSuite extends QueryTest with SQLTestUtils {
       }
     }
   }
+
+  test("SPARK-41982: When the partition is not enclosed with quotation marks, " +
+    "the partition type is string and keepPartitionSpecAsStringLiteral is enabled, " +
+    "we need to treat it as a string type partition") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLInsertTestSuite.scala:357]: can we shorten the test name?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala (hunk 3 above)
Comment (smallzhongfeng) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlParser.scala:366]: Good idea!

File: sql/core/src/test/scala/org/apache/spark/sql/SQLInsertTestSuite.scala (hunk 4 above)
Comment (smallzhongfeng) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLInsertTestSuite.scala]: Updated.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 2 above)
Comment (smallzhongfeng) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Yes.

**Comment by smallzhongfeng** [comment]
Could you help me review again ? @cloud-fan

File: docs/sql-migration-guide.md [hunk 5]
```diff
@@ -36,6 +36,7 @@ license: |
...
   - Since Spark 3.4, Spark will do validation for partition spec in ALTER PARTITION to follow the behavior of `spark.sql.storeAssignmentPolicy` which may cause an exception if type conversion fails, e.g. `ALTER TABLE .. ADD PARTITION(p='a')` if column `p` is int type. To restore the legacy behavior, set `spark.sql.legacy.skipTypeValidationOnAlterPartition` to `true`.
   - Since Spark 3.4, vectorized readers are enabled by default for the nested data types (array, map and struct). To restore the legacy behavior, set `spark.sql.orc.enableNestedColumnVectorizedReader` and `spark.sql.parquet.enableNestedColumnVectorizedReader` to `false`.
+  - Since Spark 3.4, When inserting(adding/renaming/dropping) a partition, the partition field is of `String` and the partition value is not quoted, it will be treated as a field of `Numeric` by default. If you want to keep the partition field of `String`, we can set `spark.sql.legacy.keepPartitionSpecAsStringLiteral` to `true`.
```
Comment (cloud-fan) [inline docs/sql-migration-guide.md:39]: We already added the migration guide when introducing the breaking change, no need to add it again.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 6]
```diff
@@ -3288,6 +3288,15 @@ object SQLConf {
       .booleanConf
       .createWithDefault(false)
 
+  val LEGACY_KEEP_PARTITION_SPEC_AS_STRING_LITERAL =
+    buildConf("spark.sql.legacy.keepPartitionSpecAsStringLiteral")
+      .internal()
+      .doc("If it is set to true and the type of the partition is string, " +
+        "partition value will be treated as string value but not numeric value")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3295]: can you follow the config doc above? I think it's clearer.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLInsertTestSuite.scala [hunk 7]
```diff
@@ -351,6 +351,35 @@ trait SQLInsertTestSuite extends QueryTest with SQLTestUtils {
       }
     }
   }
+
+  test("SPARK-41982: " +
+    "treat the partition field as a string type when keepPartitionSpecAsStringLiteral is enabled") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLInsertTestSuite.scala:356]:
```suggestion
    "treat the partition field as string literal when keepPartitionSpecAsStringLiteral is enabled") {
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 6 above)
Comment (smallzhongfeng) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3295]: Updated.

**Comment by smallzhongfeng** [comment]
All comments have been addressed, PTAL, thanks @cloud-fan

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by smallzhongfeng** [comment]
> thanks, merging to master!

Thanks for your review! @cloud-fan

**Comment by gatorsmile** [comment]
@smallzhongfeng Could you help add it to the migration guide?

**Comment by smallzhongfeng** [comment]
Sure, but the result of the previous discussion is that there is no need to add, you can see https://github.com/apache/spark/pull/39558#discussion_r1071207116 I will add it if necessary. @gatorsmile

**Pull request closed without merging**